Make sct_get_centerline robust to intensities ∈ [0, 1]

charleygros opened the pull request:
### Requirements

* [x] Is there a issue open describing the motivation for this pull request ?
https://github.com/neuropoly/spinalcordtoolbox/issues/1740

### Description of the Change
If all intensities of the image are < 1, then this line:
https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/centerline/optic.py#L95
leads to a binarization of the input image --> fail of OptiC

Solution: make an intensity scaling prior to the int16 conversion.
Changes tested on ~ 50 images --> seems to do not affect OptiC.

### Steps and Constraints

To test the pull request, please reinstall the toolbox because the changes are made in spinalcordtoolbox/optic

jcohenadad commented on spinalcordtoolbox/centerline/optic.py:
```diff
@@ -63,6 +64,17 @@ def centerline2roi(fname_image, folder_output='./', verbose=0):
+def apply_intensity_normalization(fname_in, fname_out):
+    img = Image(fname_in)
+    img_normalized = img.copy()
+    p2, p98 = np.percentile(img.data, (2, 98))
```
Isn't int16 range -32,768 to 32,767?

charleygros replied:
I am sorry, I am not sure to understand: does the below comment helps? Otherwise, May I ask you to re-explain to me? Thank you

charleygros replied:
```
Wouldn't it be better to fix the issue inside sct_image instead?
```
Maybe! But I am not sure to see how, Could you please help me? Merci

jcohenadad replied:
sorry @charleygros i meant line 71 (below), where it says:  out_range=(0, 255) (equivalent 8bit unsigned), whereas later in the code the data is converted to int16, hence we loose precision. Or am i missing something?

jcohenadad replied:
maybe [skimage.img_as_uint](http://scikit-image.org/docs/dev/api/skimage.html#skimage.img_as_uint) could be useful here?

charleygros replied:
It seems to be a better solution, merci!

Just one question: The documentation says:
```
Negative input values will be clipped.
```
--> Could we assume that images will always have positive values?

jcohenadad replied:
i suggest we do, but we should check for it. If there are neg values, a "warning" should be sent.

charleygros commented:
The issue here is:
```
cd /Volumes/sct_testing/issues/20180517_issue1740
sct_image -i 9611-0_0-images_017_jvgre10TE2mm1001_trace15Cav.nii.gz -type int16 -o 9611-0_0-images_017_jvgre10TE2mm1001_trace15Cav_int16.nii.gz
```
Leads to:
![screen shot 2018-05-17 at 8 15 38 pm](https://user-images.githubusercontent.com/14353425/40210431-eaba4596-5a12-11e8-90cd-3a556afb4b0b.png)


Concerning the -32,768 to 32,767:
```
cd /Volumes/sct_testing/issues/20180517_issue1740
python
>>> from msct_image import Image
>>> import numpy as np
>>> ii = Image('9611-0_0-images_017_jvgre10TE2mm1001_trace15Cav_-31_689_31_115999.nii.gz')
>>> np.min(ii.data), np.max(ii.data)
(-31.698, 31.115999)
>>> del ii
>>> quit()
sct_get_centerline -i 9611-0_0-images_017_jvgre10TE2mm1001_trace15Cav_-31_689_31_115999.nii.gz -c t2s
```
Leads to:
![screen shot 2018-05-17 at 8 41 46 pm](https://user-images.githubusercontent.com/14353425/40210412-c826bb40-5a12-11e8-93fd-9c89443ab733.png)

jcohenadad commented:
> Wouldn't it be better to fix the issue inside sct_image instead?

I will look into it and provide suggestions

jcohenadad commented:
in response to [this comment](https://github.com/neuropoly/spinalcordtoolbox/pull/1746#issuecomment-390065138): can't we do a rescaling, by adjusting the in_min and in_max to the out_min and out_max, as done [here](https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/msct_image.py#L392)?

charleygros commented:
Travis showed error of `sct_testing -f sct_propseg`.
Reason:
https://github.com/neuropoly/spinalcordtoolbox/blob/master/scripts/sct_propseg.py#L530
The output of PropSeg is converted to `int8` --> a scaling of [0, 1] leads to [-127, 128] --> Dice = 0, because of this commit:
https://github.com/neuropoly/spinalcordtoolbox/pull/1746/commits/f26a73513ccb6a7037ca82d29089646de9f88d3a

I would suggest to:
- keep the former implementation of msct_image.changeType since removing the "if condition" will certainly break other functions
- use the `skimage.img_as_uint` in `optic.py` combined with: if negative values --> apply a simple translation to make them positive.

jcohenadad commented:
Let's do a comparative test as in #1757 before merging. I'll take care of that and add results in this PR.

charleygros commented:
@jcohenadad: By comparing the two pickles (master and cg_i1740):
```
path_cg = '/Users/chgroc/code/i1740/cg_i.pickle'
path_master = '/Users/chgroc/code/i1740/master.pickle'

pickle_file = open(path_cg)
cg_data = pickle.load(pickle_file)
cg_status99 = cg_data[(cg_data.status == 99)]['path_data'].apply(lambda x: x.split('/')[5])

pickle_file = open(path_master)
master_data = pickle.load(pickle_file)
master_status99 = master_data[(master_data.status == 99)]['path_data'].apply(lambda x: x.split('/')[6])

print list(set(set(cg_status99)-set(master_status99)))
```
we obtain the following result:
```
['milan_20160628-marcella_DCD-20110907']
```

I will investigate further this data
